Frontend/Frontend.Client/Models/WebSocketService.cs

```diff
@@ -0,0 +1,43 @@
+﻿namespace Frontend.Client.Models;
+
+using System.Net.WebSockets;
+using System.Text;
+
+public class WebSocketService
+{
+    private ClientWebSocket? _webSocket;
+    public event Action<string>? OnMessageReceived;
+
+    public async Task ConnectAsync(string uri)
+    {
+        if (_webSocket == null || _webSocket.State != WebSocketState.Open)
+        {
+            _webSocket = new ClientWebSocket();
+            await _webSocket.ConnectAsync(new Uri(uri), CancellationToken.None);
+            _ = ReceiveMessagesAsync();
+        }
+    }
+
+    private async Task ReceiveMessagesAsync()
+    {
+        var buffer = new byte[1024 * 4];
+        while (_webSocket?.State == WebSocketState.Open)
+        {
+            var result = await _webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);
+            if (result.MessageType == WebSocketMessageType.Text)
+            {
+                var message = Encoding.UTF8.GetString(buffer, 0, result.Count);
+                OnMessageReceived?.Invoke(message);
```

Comment:
Preferrably can we call an OnMessageReceive function here rather than trigger an event? That way we can keep all the OnMessage implementation in one place rather than different areas subscribing to the event which can be chaotic and could lead to race conditions.

Reply:
idk what u mean by race conditions tbh